### DISCUSSION TITLE
Support custom events

### DISCUSSION THREAD

**Discussion by asonix** [davidpdrsn/axum-live-view#77] [Q&A] [answered]
I'm looking through the examples right now trying to wrap my head around events. It seems like events are currently limited to static string keys, but it would be interesting to be able to embed more data in each event.

e.g.

```rust
enum MyEvent {
    MyTextInputMessage {
        id: Uuid,
        contents: String,
    },
}

impl LiveView for MyHandler {
    fn setup(&self, sub: &mut Setup<Self>) {
        sub.on_custom(MyHandler::custom);
    }
}

impl MyHandler {
    fn custom(mut self, msg: MyEvent) -> Self {
        ...
    }
}
```

```rust
for item in some_collection {
    <input type="text" { CustomInput }={ |input_text| MyEvent::MyTextInputMessage { id: item.id, contents: input_text } } />
}
```

**Comment by davidpdrsn** [comment, marked as answer]
This is already supported actually, its just not clear from the the examples.

You can do this:

```rust
#[derive(Default)]
struct View;

impl LiveView for View {
    fn setup(&self, subscriptions: &mut Setup<Self>) {
        subscriptions.on("do-thing", Self::do_thing);
    }

    fn render(&self) -> Html {
        html! {
            <button
                { Click }={ "do-thing" }
                // `axm_data` is a convenience for creating an `axm-data-` attribute
                // its probably wont be the final name. I would like `Axm::Data(...)` but that requires
                // some changes to the `Axm` enum
                { axm_data("one") }="foo"
                // this works too
                axm-data-two="bar"
                // The values can also be dynamic
                { axm_data("three") }={ 1 + 2 }
            >"Do thing!"</button>
        }
    }
}

impl View {
    async fn do_thing(self, axum::Json(msg): axum::Json<CustomMessage>) -> Self {
        dbg!(&msg);
        self
    }
}

#[derive(Deserialize, Debug)]
struct CustomMessage {
    one: String,
    two: String,
    three: String,
}
```

So all `axm-data-{...}` attributes will be send along with events from that element. The names after `axm-data-` will become field names mapped to deserializable structs.

All this will of course be documented at some point 😊

**Reply by asonix** [reply]
That's really neat! although it does seem to require that the Message type implement `serde::Deserialize`. My idea with the `|input_text| -> Message` callback was to avoid serializing/deserializing data entirely. e.g. the `input_text` is the only data sent from the client, and the callback is executed on the server to produce the custom message. An implementation would likely need to uniquely tag all custom inputs like this to associate the message-builder Fn with the text produced like a `text_inputs: HashMap<Id, Box<dyn Fn(String) -> Msg>>` where the HashMap is produced as a side-effect of the render

My goal with this was to enable arbitrary custom message types, so I could build a (potentially nested) message enum similar to a `Msg` custom type in an Elm application: https://github.com/rtfeldman/elm-spa-example/blob/cb32acd73c3d346d0064e7923049867d8ce67193/src/Main.elm#L114-L124

roughly translated:
```rust
enum Msg {
    ChangedUrl(Url),
    ClickedLink(browser::UrlRequest),
    GotHomeMsg(home::Msg),
    GotLoginMsg(login::Msg),
    ...
}
```

Messages like these are not only used for _consuming_ events, but also for _producing_ events, like so: https://github.com/rtfeldman/elm-spa-example/blob/cb32acd73c3d346d0064e7923049867d8ce67193/src/Page/Login.elm#L137

roughly translated:
```rust
fn update(mut self, msg: Msg) -> Self {
    match msg {
        ...
        Msg::GotLoginMsg(login_msg) => Self {
             login = self.login.update(login_msg),
             ..self
        }
        ...
    }
}

fn render(&self) -> Html<Msg> {
    ...
    self.login.render().map(Msg::GotLoginMsg)
    ...
}

mod login {
    enum Msg {
        EnteredEmail(String),
        EnteredPassword(String),
    }
    
    fn update(mut self, msg: Msg) -> Self {
        ...
    }
    
    fn render(&self) -> Html<Msg>
        html! {
            <input { CustomInput }={ |s| Msg::EnteredPassword(s) } />
            <input { CustomInput }={ Msg::EnteredEmail } /> // or shorthand type constructor as Fn
        }
    }
}
```

To accomplish something similar with the current api, an input would need to be constructed like so:
```rust
#[derive(serde::Deserialize)]
#[serde(tag = "page")]
enum Msg {
    #[serde(flatten)]
    Login(login::Msg),
}

fn on_text_update(self, form_event: FormEvent, Json(mut extra): Json<serde_json::Value>) -> Self {
    extra["value"] = form_event.into_value();

    if let Ok(msg) = serde_json::from_value(extra) {
        return self.update(msg);
    }
    
    self
}

fn update(mut self, msg: Msg) -> Self {
    match msg {
        ...
        Msg::Login(login_msg) => Self {
            login: self.login.update(login_msg),
            ..self
        }
        ...
    }
}

fn render(&self) -> Html {
    ...
    self.login.render()
    ...
}

mod login {
    #[derive(serde::Deserialize)]
    #[serde(tag = "kind")]
    enum Msg {
        #[serde(flatten)]
        Email {
            value: String,
        },
        #[serde(flatten)]
        Password {
            value: String,
        },
    }
    
    fn update(mut self, msg: Msg) -> Self {
        ...
    }
    
    fn render(&self) -> Html {
        html! {
            <input { Input }="text-input" { axm_data("page") }="login" { axm_data("kind") }="email" />
            <input { Input }="text-input" { axm_data("page") }="login" { axm_data("kind") }="password" />
        }
    }
}
```

Needing to specify struct fields like this severely limits the shapes the `Message` type can have, and makes nesting messages hazardous due to potentially conflicting field names, needing to keep static strings consistent with enum variants, adding all the right serde annotations, etc.

**Reply by davidpdrsn** [reply]
Right I see what you mean but I don’t think we’ll support something like the “elm architecture”, at least at first. I find the current model works quite well and I’m not yet convinced either is objectively better.

Nested views would be good to support in some form or another. Tracked in https://github.com/davidpdrsn/axum-liveview/issues/61. Haven’t thought much about it yet though.

**Reply by davidpdrsn** [reply]
And there is no escaping serialization since if you have multiple server replicas the view and the socket might be on different replicas. So they need a way to communicate. That’s why “pubsub” exists.

**Reply by asonix** [reply]
I'm confused about that, are you saying the the task that's spawned to manage the websocket connection might live on a different server than the one that calls `render`?

**Reply by davidpdrsn** [reply]
Yes

**Reply by davidpdrsn** [reply]
The server that receives the initial request and the websocket connect request might not be the same if you’re load balancing multiple servers.

**Reply by asonix** [reply]
So why doesn't the `LiveView` trait require that `Self: Encode + Decode`?

**Reply by davidpdrsn** [reply]
Where do you see that? https://github.com/davidpdrsn/axum-liveview/blob/main/axum-liveview/src/liveview.rs

**Reply by asonix** [reply]
I'm asking why it _doesn't_, the `LiveView` type clearly exists on a single server, and is never serialized and sent between servers. Since this is the case, the `Message` type also does not need to be serialized or deserialized, it can be constructed from the required parts in the same task that the `LiveView` type exists.

**Reply by davidpdrsn** [reply]
I don't understand what you mean. As things work today, the live view is created with the initial request. Once that's done the browser opens a web socket connection that needs to talk to the view that's running in an async task somewhere. The websocket request has no idea what views are actually on the page. Only the initial request knows since it has the initial html the views are embedded in.

Requiring live views to be encodeable would be very limiting. How do you encode a database connection?

**Reply by asonix** [reply]
nonono I'm not asking that we make `LiveView: Encode + Decode`. I'm saying that _because_ `LiveView` is _not_ `Encode + Decode` I know that a LiveView never moves between servers (And that's fine! probably good, even)

And because I know that a LiveView never moves between servers, I know that associated state can be kept with a LiveView in its task, like a `HashMap<Id, Box<dyn Fn(String) -> Msg>>`. HashMaps like this can be responsible for producing `Msg` types that stay within the LiveView's task, but are easier for developers building LiveViews to reason about. Developers can include their "Higher Level" Message type into their HTML, but that type never needs to leave the LiveView task because it can be constructed from lower-level pieces (like the string somebody has typed into a text input).

Basically after a call to `render(&self)` that produces an `Html<Msg>`, there can be an additional step that does the following:
```rust
fn render(state: &mut SomeState, live_view: &LiveView) -> PlainHtml {
   let rendered: Html<Msg> = live_view.render();
   
   let (html, callbacks): (PlainHtml, Callbacks<Msg>) = rendered.split();
   
   state.register_callbacks(callbacks);
   
   html
}
```

**Reply by asonix** [reply]
perhaps I could build this on top of how the LiveView type already works 🤔 , well except that I don't think I can produce my own `Html` type without that macro

**Reply by davidpdrsn** [reply]
Sorry but I'm still not following. Are you proposing a new feature or architecture? If so, what problems do they address? Or are you just asking questions (which is fine)? It's a bit late here and I'm behind on sleep, so I'm probably just tired 😅

My opinion is still that the elm architecture (init, update, render, and one big message enum) isn't necessarily better than what we have now, outside of natural nesting of components. But that has other solutions I'm sure.

I did consider making `Html` generic over the messages it sends so `Html<YourMessageType>`, like Elm does. But I don't know how that would work for messages that don't come from the view. Currently any part of the application that has an `impl PubSub` can send messages to any anything that happens to be subscribed.

**Reply by asonix** [reply]
I'm just trying to build Elm on top of axum-liveview tbh

I think something that could unblock people from experimenting with things like this outside of the axum-liveview repository is if there was a way to construct `Html` values without using the `html!` macro. For example, I could build my own `Html<Msg>` type that I declare in my own LiveView type that I would internally translate to an `Html`

**Reply by davidpdrsn** [reply]
The `Html` type is very closely tied to diffing the views so we can compute minimal deltas. 

Allowing users to create their own without using the html macro is technically possible but probably not a goal at first.

If you expand the macro you'll see it looks quite different from the code you write 😅

But I do actually have a few thoughts on how to make messages more typesafe. That's an area that's currently lacking. I'm gonna do some experiments and might ping you on PRs if you're up for giving feedback?

**Reply by asonix** [reply]
sounds great!

**Comment by davidpdrsn** [comment]
And cool that you're already checking out axum-liveview! Do you happen to have any name suggestions? I don't actually want to use "liveview" since thats already a thing and bad for googling.

**Reply by asonix** [reply]
It's trickier to come up with a response to this comment than the previous one :sweat_smile:

**Reply by asonix** [reply]
Maybe "LiveApp" or "LivePage"?

**Reply by davidpdrsn** [reply]
> It's trickier to come up with a response to this comment than the previous one :sweat_smile: 

Hehe yes. Naming is the hardest problem.

> Maybe "LiveApp" or "LivePage"?

Hm LiveApp isn't too bad 🤔

**Reply by asonix** [reply]
I'd hesitate to call it a "LiveApp" until it uses the browser's history API for routing instead of `window.location.href = uri;` tbh

**Reply by davidpdrsn** [reply]
Good point 👍
